Update Maven usage in README to suggest annotationProcessorPaths

tbroyer opened the pull request:
Note that the online doc (`gh-pages` ) needs to be updated too; I'll do a followup pull request if this one's accepted.

JakeWharton commented on compiler/src/it/functional-tests/pom.xml:
```diff
@@ -38,26 +38,16 @@ limitations under the License.
       <artifactId>auto-value</artifactId>
       <version>${auto.value.version}</version>
-      <scope>provided</scope> <!-- to leave out of the all-deps jar -->
+      <scope>provided</scope> <!-- only needed for the annotations -->
```
Hopefully this will motivate https://github.com/google/auto/issues/268 :grinning:

tbroyer commented:
Hmm, maybe samples should not be updated to `annotationProcessorPaths`, it has side-effects on the execution graph of the reactor build (i.e. `mvn package -pl examples/simple -am` will **not** build the annotation processor but will instead resolve it from the repositories).
See http://blog.ltgt.net/maven-is-broken-by-design-2/ for details.

It's also worth waiting for version 3.6 of the maven-compiler-plugin that will fix the incremental build issue introduced in 3.2 (three years ago or so).

Feel free to close this PR following those findings.

tbroyer commented:
Updated to only use `annotationProcessorPaths` in integration tests (not in samples), and fixed documentation in README.

Use of `annotationProcessorPaths` (i.e. `-processorpath`) can be particularly useful when using GWT, given #228, to isolate `dagger-compiler` from your classpath and prevent clashes with GWT's own bundled Eclipse JDT (which I haven't experienced, but could still be possible; I think `dagger-compiler` came after `gwt-dev` in the classpath, and `google-java-format` seems to work OK with the version of Eclipse JDT shipped within GWT; the reverse might not be true though: GWT ships a modified version of JDT with a backported patch to workaround a memory consumption issue)

ronshapiro commented on compiler/src/it/producers-functional-tests/pom.xml:
```diff
@@ -69,10 +63,17 @@ limitations under the License.
       <plugin>
         <artifactId>maven-compiler-plugin</artifactId>
-        <version>3.1</version>
+        <version>3.5.1</version>
```
Can you define a variable for this in the root `pom.xml`?

ronshapiro replied:
Or, is there a way to set the default and then have that be implicit?

ronshapiro commented on README.md:
```diff
@@ -66,6 +66,37 @@ of your `pom.xml` and the `dagger-compiler` artifact as either an `optional` or
+  <dependency>
+    <groupId>com.google.dagger</groupId>
+    <artifactId>dagger</artifactId>
+    <version>2.0.2</version>
```
Can you leave this as `2.x` so we don't need to constantly update it? We updated this after you sent out this PR.

ronshapiro commented on README.md:
````diff
@@ -66,6 +66,37 @@ of your `pom.xml` and the `dagger-compiler` artifact as either an `optional` or
 </dependencies>
 ```
 
+or as an `annotationProcessorPaths` value of the `maven-compiler-plugin`
````
Is there any reason why we shouldn't just recommend this outright?

tbroyer replied:
* requires `maven-compiler-plugin` version 3.5+
* things like `mvn dependency:go-offline`, `mvn versions:display-dependency-updates` won't see them (of course, if you use a property for the Dagger version, `mvn versions:display-property-updates` would see it from the `com.google.dagger:dagger` _standard_ dependency)
* doesn't respect `dependencyManagement` (for managing versions), forcing you to use a property. In other words, you have `<version>` scattered all over your POMs, rather than centralized in a `dependencyManagement`; and it prevents you from using BOMs (`<scope>import</scope>`) to manage versions of your dependencies across projects of your organization (forces you to use a `<parent>` defining a property for the Dagger version).

ronshapiro commented:
I'm in the process of switching us to use `bazel`, so besides the README and users guide, this won't be relevant soon (I hope). Do you still think it's useful in those two files?

tbroyer commented:
Yay for bazel though I wonder how you'll deploy releases (and snapshots?) to Central.

But yes, probably not worth it. Feel free to close.

JakeWharton commented:
You can still use `mvn` to deploy with `mvn deploy:deploy-file` ([example here][1]). You can also do individual HTTP PUTs or a multi-part POST into a Sonatype staging repo (which is all Maven does).

 [1]: https://github.com/JakeWharton/sdk-artifact-sync/blob/f90616ecc183bafc8d2c6b26a9c5945471a77846/sdk-artifact-sync.py#L182-L193

tbroyer commented:
I rather meant the non/less-technical part: ensuring java_library↔pom parity, making sure things work for non-bazel users, etc.

tbroyer commented:
Now that the build has moved to Bazel, this PR only updates the README (and I'll send a PR for th users-guide if/when that one's accepted)

ronshapiro commented:
LGTM. Let's put both the README and users guide changes in the same PR and then I'll grab all the changes in?

ronshapiro commented:
Btw don't worry about squashing since it effectively happens when we merge internally

tbroyer commented:
> Let's put both the README and users guide changes in the same PR and then I'll grab all the changes in?

users guide is in a different branch ;-)

BTW, there are Gradle instructions in the README, should those be added to the users guide too? (as we're at it)

ronshapiro commented:
I've always found it a little weird that we have this duplicated. I think it's probably in the best interest of our users though to have it in both locations. SGTM. 

If you create both PRs, I'll take a look at both and submit. I forgot that they were different branches on github

tbroyer commented:
See https://github.com/google/dagger/pull/719

The users guide was already deferring to the README for Gradle, so I removed the Maven instructions. Let me know if you'd prefer to instead duplicate all instructions in the users guide.
